RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByCandidateRegion.cc

```diff
@@ -0,0 +1,304 @@
+          }
+
+          std::vector<float> etaMin;
+          etaMin.reserve(maxNRegions_);
+          std::vector<float> etaMax;
+          etaMax.reserve(maxNRegions_);
+          std::vector<float> phiMin;
+          phiMin.reserve(maxNRegions_);
+          std::vector<float> phiMax;
+          phiMax.reserve(maxNRegions_);
+          std::vector<math::XYZPoint> origin;
+          origin.reserve(maxNRegions_);
+          std::vector<float> zBound;
+          zBound.reserve(maxNRegions_);
+
+          size_t n_objects = objectsHandle->size();
+          size_t n_regions = 0;
+          for (size_t i = 0; i < n_objects && n_regions < maxNRegions_; ++i) {
+            //	      std::cout << "i: " << i << " n_regions: " << n_regions << std::endl;
+            const reco::Candidate& object = (*objectsHandle)[i];
+            //	  GlobalVector direction( object.momentum().x(), object.momentum().y(), object.momentum().z() );
+            for (size_t j = 0; j < origins.size() && n_regions < maxNRegions_; ++j) {
+              etaMin.push_back(object.eta() - deltaEta_);
+              etaMax.push_back(object.eta() + deltaEta_);
+              phiMin.push_back(Geom::Phi(object.phi() - deltaPhi_));
+              phiMax.push_back(Geom::Phi(object.phi() + deltaPhi_));
+              origin.push_back(origins[j].first);
+              zBound.push_back(origins[j].second);
+
+              n_regions++;
+            }
+          }
+
+          size_t it = 0;
+          const auto curvature1 = (value1_ >= 0.) ? PixelRecoUtilities::curvature(1.f / pt1_, eSetup) : -1.;
+          const auto curvature2 = (value2_ >= 0.) ? PixelRecoUtilities::curvature(1.f / pt2_, eSetup) : -1.;
+          for (auto trk : tracks) {
+            const auto pt = trk.pt();
+            if (pt < minPt_)
+              continue;
+            if (value1_ >= 0.) {
+              const auto curvature = PixelRecoUtilities::curvature(1.f / pt, eSetup);
+              // https://cmssdt.cern.ch/dxr/CMSSW/source/RecoPixelVertexing/PixelTriplets/interface/CAHitQuadrupletGenerator.h#58
+              auto value = -1.;
+              if (curvature1 < curvature)
+                value = value1_;
+              else if (curvature2 < curvature && curvature <= curvature1)
+                value = value2_ + (curvature - curvature2) / (curvature1 - curvature2) * (value1_ - value2_);
+              else
+                value = value2_;
+
+              const auto chi2 = trk.chi2();
+              if (chi2 > value)
+                continue;
+            }
+
+            const auto eta = trk.eta();
+            const auto phi = trk.phi();
+
+            for (size_t k = 0; k < n_regions; ++k) {
+              if (fabs(trk.dz(origin[k])) > zBound[k])
+                continue;
+              //		if ( fabs(trk.vz() - origin[k].z()) > zBound[k] ) continue;
+
+              if ((eta >= etaMin[k] && eta <= etaMax[k]) && (phi >= phiMin[k] && phi <= phiMax[k])) {
```

Comment:
This doesn't appear to handle the region near +/-pi appropriately
IIUC, e.g. for the region pi-0.05 and dphi 0.1 we get phiMin = pi-0.15, phiMax = -pi + 0.05 and "phi" here is a simple double and that way nothing will pass